Optimize grouped conv2d by unfold -> bmm

piyushmh opened the pull request:
Summary: The current implementation of group conv2d loops over group. This change batches all the groups for a more efficient computation. 

* supporting grouped conv2d in slow_conv2d
* adding a fast path in __convolution to call slow_conv2d when running grouped_conv2d on CPU

Test Plan:
Added the following test case  in test_nn.py, testing both forward and backward pass:
* test_Conv2d_groups_nobias_mklnn_off

```
PyTorch built with:
  - GCC 4.2
  - C++ Version: 201703
  - clang 8.0.20181009
  - Intel(R) Math Kernel Library Version 2017.0.3 Product Build 20170413 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v1.2.1 (Git Hash N/A)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - CPU capability usage: AVX2
  - Build settings: 
```

```
from timeit import Timer

import numpy as np
import torch


num = 50
S = [
    [1, 1024, 14, 14, 512, 1, 1, 0, 0],
    [1, 1024, 14, 14, 512, 3, 3, 0, 0],
    [1, 1024, 14, 14, 512, 5, 5, 0, 0],
]

for g in [2, 4, 8]:
    for padding in [0, 1]:
        for x in range(len(S)):
            P = S[x]
            (N, C, H, W) = P[0:4]
            M = P[4]
            (kernel_h, kernel_w) = P[5:7]
            (stride_h, stride_w) = P[7:9]
            (padding_h, padding_w) = (padding, padding)

            X_np = np.random.randn(N, C, H, W).astype(np.float32)
            W_np = np.random.randn(M, C, kernel_h, kernel_w).astype(np.float32)
            X = torch.from_numpy(X_np)
            print(
                (
                    "N: {N}, C: {C}, H: {H}, W: {W}, M: {M}, kernel_h: {kernel_h}, kernel_w: {kernel_w},"
                    "stride_h: {stride_h}, stride_w: {stride_w}, padding_h: {padding_h}, "
                    "padding_w: {padding_w} groups: {groups}"
                ).format(
                    N=N,
                    C=C,
                    H=H,
                    W=W,
                    M=M,
                    kernel_h=kernel_h,
                    kernel_w=kernel_w,
                    stride_h=stride_h,
                    stride_w=stride_w,
                    padding_h=padding_h,
                    padding_w=padding_w,
                    groups=g,
                )
            )
            conv2d_pt = torch.nn.Conv2d(
                C,
                M,
                (kernel_h, kernel_w),
                stride=(stride_h, stride_w),
                padding=(padding_h, padding_w),
                groups=g,
                bias=True,
            )

            class ConvNet(torch.nn.Module):
                def __init__(self):
                    super(ConvNet, self).__init__()
                    self.conv2d = conv2d_pt

                def forward(self, x):
                    return self.conv2d(x)

            model = ConvNet()

            def pt_forward():
                with torch.no_grad():
                    model(X)

            torch._C._set_mkldnn_enabled(True)
            t = Timer("pt_forward()", "from __main__ import pt_forward, X")
            print("MKLDNN pt time = {}".format(t.timeit(num) / num * 1000.0))
            torch._C._set_mkldnn_enabled(False)
            t = Timer("pt_forward()", "from __main__ import pt_forward, X")
            print("TH pt time = {}".format(t.timeit(num) / num * 1000.0))
            print("\n")
```
Command:
OMP_NUM_THREADS=1 MKL_NUM_THREADS=1 buck run @mode/opt experimental/piyushmh:group_conv2d_perf

Results:
```
N: 1, C: 1024, H: 14, W: 14, M: 512, kernel_h: 1, kernel_w: 1,stride_h: 1, stride_w: 1, padding_h: 0, padding_w: 0 groups: 2
MKLDNN pt time = 2.396249941084534
TH pt time = 1.6212143609300256
New implementation TH pt time = 4.276635500136763


N: 1, C: 1024, H: 14, W: 14, M: 512, kernel_h: 3, kernel_w: 3,stride_h: 1, stride_w: 1, padding_h: 0, padding_w: 0 groups: 2
MKLDNN pt time = 14.524029481690377
TH pt time = 6.749001618009061
New implementation TH pt time = 6.981048060115427

N: 1, C: 1024, H: 14, W: 14, M: 512, kernel_h: 5, kernel_w: 5,stride_h: 1, stride_w: 1, padding_h: 0, padding_w: 0 groups: 2
MKLDNN pt time = 33.35205713985488
TH pt time = 17.04349328065291
New implementation TH pt time = 23.01782624097541

N: 1, C: 1024, H: 14, W: 14, M: 512, kernel_h: 1, kernel_w: 1,stride_h: 1, stride_w: 1, padding_h: 1, padding_w: 1 groups: 2
MKLDNN pt time = 2.1534182014875114
TH pt time = 1.6464285389520228
New implementation TH pt time = 1.6756533994339406


N: 1, C: 1024, H: 14, W: 14, M: 512, kernel_h: 3, kernel_w: 3,stride_h: 1, stride_w: 1, padding_h: 1, padding_w: 1 groups: 2
MKLDNN pt time = 14.301151898689568
TH pt time = 9.977071180474013
New implementation TH pt time = 11.18448490044102



N: 1, C: 1024, H: 14, W: 14, M: 512, kernel_h: 5, kernel_w: 5,stride_h: 1, stride_w: 1, padding_h: 1, padding_w: 1 groups: 2
MKLDNN pt time = 37.25171020021662
TH pt time = 25.08984894026071
New implementation TH pt time = 28.640052361879498



N: 1, C: 1024, H: 14, W: 14, M: 512, kernel_h: 1, kernel_w: 1,stride_h: 1, stride_w: 1, padding_h: 0, padding_w: 0 groups: 4
MKLDNN pt time = 1.214240558911115
TH pt time = 0.8866383391432464
New implementation TH pt time = 0.7292797393165529



N: 1, C: 1024, H: 14, W: 14, M: 512, kernel_h: 3, kernel_w: 3,stride_h: 1, stride_w: 1, padding_h: 0, padding_w: 0 groups: 4
MKLDNN pt time = 4.446479140315205
TH pt time = 3.8826309191063046
New implementation TH pt time = 3.8437813404016197


N: 1, C: 1024, H: 14, W: 14, M: 512, kernel_h: 5, kernel_w: 5,stride_h: 1, stride_w: 1, padding_h: 0, padding_w: 0 groups: 4
MKLDNN pt time = 17.763572039548308
TH pt time = 10.842112661339343
New implementation TH pt time = 14.842092178296298


N: 1, C: 1024, H: 14, W: 14, M: 512, kernel_h: 1, kernel_w: 1,stride_h: 1, stride_w: 1, padding_h: 1, padding_w: 1 groups: 4
MKLDNN pt time = 1.4290818781591952
TH pt time = 1.2468176404945552
New implementation TH pt time = 1.1342144012451172


N: 1, C: 1024, H: 14, W: 14, M: 512, kernel_h: 3, kernel_w: 3,stride_h: 1, stride_w: 1, padding_h: 1, padding_w: 1 groups: 4
MKLDNN pt time = 6.148020739201456
TH pt time = 6.651792719494551
New implementation TH pt time = 6.367272478528321


N: 1, C: 1024, H: 14, W: 14, M: 512, kernel_h: 5, kernel_w: 5,stride_h: 1, stride_w: 1, padding_h: 1, padding_w: 1 groups: 4
MKLDNN pt time = 19.79779327986762
TH pt time = 11.329268079716712
New implementation TH pt time = 20.390679300762713



N: 1, C: 1024, H: 14, W: 14, M: 512, kernel_h: 1, kernel_w: 1,stride_h: 1, stride_w: 1, padding_h: 0, padding_w: 0 groups: 8
MKLDNN pt time = 0.8378969598561525
TH pt time = 0.7001999602653086
New implementation TH pt time = 0.5394901405088603



N: 1, C: 1024, H: 14, W: 14, M: 512, kernel_h: 3, kernel_w: 3,stride_h: 1, stride_w: 1, padding_h: 0, padding_w: 0 groups: 8
MKLDNN pt time = 2.4611629196442664
TH pt time = 2.8066396596841514
New implementation TH pt time = 2.6308178016915917



N: 1, C: 1024, H: 14, W: 14, M: 512, kernel_h: 5, kernel_w: 5,stride_h: 1, stride_w: 1, padding_h: 0, padding_w: 0 groups: 8
MKLDNN pt time = 4.467485479544848
TH pt time = 6.293305980507284
New implementation TH pt time = 12.03830363927409

N: 1, C: 1024, H: 14, W: 14, M: 512, kernel_h: 1, kernel_w: 1,stride_h: 1, stride_w: 1, padding_h: 1, padding_w: 1 groups: 8
MKLDNN pt time = 1.1003097402863204
TH pt time = 1.0144947399385273
New implementation TH pt time = 0.7655130396597087



N: 1, C: 1024, H: 14, W: 14, M: 512, kernel_h: 3, kernel_w: 3,stride_h: 1, stride_w: 1, padding_h: 1, padding_w: 1 groups: 8
MKLDNN pt time = 3.037271238863468
TH pt time = 4.349836420733482
New implementation TH pt time = 3.942242639604956



N: 1, C: 1024, H: 14, W: 14, M: 512, kernel_h: 5, kernel_w: 5,stride_h: 1, stride_w: 1, padding_h: 1, padding_w: 1 groups: 8
MKLDNN pt time = 5.924453181214631
TH pt time = 8.270565180573612
New implementation TH pt time = 14.920658320188522
```

Differential Revision: D22344476

ngimel commented:
Are there any performance benchmarks? How does performance compare to mkldnn?

ngimel commented:
Test errors are real. Can you please let the reviewers know when it is ready for review, and provide some benchmarking numbers?

piyushmh commented:
Sorry I exported too soon. I'll fix the error and let the reviewers know when this is ready for review.

piyushmh commented:
Performance

thnn new implementation -
C = 1024
H = 256
W = 256
M = 512
554.057ms	https://www.internalfb.com/intern/paste/P135525809/

thnn old implementation -
C = 1024
H = 256
W = 256
M = 512
628.481ms	https://www.internalfb.com/intern/paste/P135523706/

mkldnn
C = 1024
H = 256
W = 256
M = 512
599.388ms	https://www.internalfb.com/intern/paste/P135525282/

lly-zero-one commented:
> Performance
> 
> thnn new implementation -
> C = 1024
> H = 256
> W = 256
> M = 512
> 554.057ms https://www.internalfb.com/intern/paste/P135525809/
> 
> thnn old implementation -
> C = 1024
> H = 256
> W = 256
> M = 512
> 628.481ms https://www.internalfb.com/intern/paste/P135523706/
> 
> mkldnn
> C = 1024
> H = 256
> W = 256
> M = 512
> 599.388ms https://www.internalfb.com/intern/paste/P135525282/

Can you show all the hyper parameters? Also make a group to show the speedup.

piyushmh committed:
Optimize grouped Conv2d performance (#41167)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/41167

Optimize grouped Conv2d performance

* supporting grouped conv2d in slow_conv2d
* adding a fast path in __convolution to call slow_conv2d when running grouped_conv2d on CPU

Test Plan:
Added the following test case  in test_nn.py, testing both forward and backward pass:
* test_Conv2d_groups_nobias_mklnn_off

Performance

```
from timeit import Timer

import numpy as np
import torch

num = 50
N = 1
C = 1024
H = 256
W = 256
M = 512
kernel_h = 1
kernel_w = 1
stride_h = 1
stride_w = 1
padding_h = 0
padding_w = 0

X_np = np.random.randn(N, C, H, W).astype(np.float32)
X = torch.from_numpy(X_np)

conv2d_pt = torch.nn.Conv2d(
    C,
    M,
    (kernel_h, kernel_w),
    stride=(stride_h, stride_w),
    padding=(padding_h, padding_w),
    groups=8,
    bias=False,
)

class ConvNet(torch.nn.Module):
    def __init__(self):
        super(ConvNet, self).__init__()
        self.conv2d = conv2d_pt

    def forward(self, x):
        return self.conv2d(x)

model = ConvNet()

def pt_forward():
    with torch.no_grad():
        model(X)

if __name__ == "__main__":
    torch._C._set_mkldnn_enabled(False)
    t = Timer("pt_forward()", "from __main__ import pt_forward, X")
    print("THH pt time = {}".format(t.timeit(num) / num * 1000.0))
    torch._C._set_mkldnn_enabled(True)
    t = Timer("pt_forward()", "from __main__ import pt_forward, X")
    print("MKLDNN pt time = {}".format(t.timeit(num) / num * 1000.0))

```

OMP_NUM_THREADS=1 MKL_NUM_THREADS=1 buck run mode/opt experimental/piyushmh:group_conv2d_perf

THH pt time = 556.7230367194861
MKLDNN pt time = 658.9539455785416

Differential Revision: D22344476

fbshipit-source-id: 4324efef0d9c83f3335b2172e4b064cd34658d75

piyushmh commented:
Here is the full script I used to test - 

```
from timeit import Timer

import numpy as np
import torch

num = 50
N = 1
C = 1024
H = 256
W = 256
M = 512
kernel_h = 1
kernel_w = 1
stride_h = 1
stride_w = 1
padding_h = 0
padding_w = 0

X_np = np.random.randn(N, C, H, W).astype(np.float32)
X = torch.from_numpy(X_np)

conv2d_pt = torch.nn.Conv2d(
    C,
    M,
    (kernel_h, kernel_w),
    stride=(stride_h, stride_w),
    padding=(padding_h, padding_w),
    groups=8,
    bias=False,
)


class ConvNet(torch.nn.Module):
    def __init__(self):
        super(ConvNet, self).__init__()
        self.conv2d = conv2d_pt

    def forward(self, x):
        return self.conv2d(x)


model = ConvNet()


def pt_forward():
    with torch.no_grad():
        model(X)


if __name__ == "__main__":
    torch._C._set_mkldnn_enabled(False)
    t = Timer("pt_forward()", "from __main__ import pt_forward, X")
    print("THH pt time = {}".format(t.timeit(num) / num * 1000.0))
    torch._C._set_mkldnn_enabled(True)
    t = Timer("pt_forward()", "from __main__ import pt_forward, X")
    print("MKLDNN pt time = {}".format(t.timeit(num) / num * 1000.0))
```

$ OMP_NUM_THREADS=1 MKL_NUM_THREADS=1 buck run @mode/opt experimental/piyushmh:group_conv2d_perf

Output - 

THH pt time = 556.7230367194861
MKLDNN pt time = 658.9539455785416

BIT-silence commented:
> Here is the full script I used to test -
> 
> ```
> from timeit import Timer
> 
> import numpy as np
> import torch
> 
> num = 50
> N = 1
> C = 1024
> H = 256
> W = 256
> M = 512
> kernel_h = 1
> kernel_w = 1
> stride_h = 1
> stride_w = 1
> padding_h = 0
> padding_w = 0
> 
> X_np = np.random.randn(N, C, H, W).astype(np.float32)
> X = torch.from_numpy(X_np)
> 
> conv2d_pt = torch.nn.Conv2d(
>     C,
>     M,
>     (kernel_h, kernel_w),
>     stride=(stride_h, stride_w),
>     padding=(padding_h, padding_w),
>     groups=8,
>     bias=False,
> )
> 
> 
> class ConvNet(torch.nn.Module):
>     def __init__(self):
>         super(ConvNet, self).__init__()
>         self.conv2d = conv2d_pt
> 
>     def forward(self, x):
>         return self.conv2d(x)
> 
> 
> model = ConvNet()
> 
> 
> def pt_forward():
>     with torch.no_grad():
>         model(X)
> 
> 
> if __name__ == "__main__":
>     torch._C._set_mkldnn_enabled(False)
>     t = Timer("pt_forward()", "from __main__ import pt_forward, X")
>     print("THH pt time = {}".format(t.timeit(num) / num * 1000.0))
>     torch._C._set_mkldnn_enabled(True)
>     t = Timer("pt_forward()", "from __main__ import pt_forward, X")
>     print("MKLDNN pt time = {}".format(t.timeit(num) / num * 1000.0))
> ```
> 
> $ OMP_NUM_THREADS=1 MKL_NUM_THREADS=1 buck run @mode/opt experimental/piyushmh:group_conv2d_perf
> 
> Output -
> 
> THH pt time = 556.7230367194861
> MKLDNN pt time = 658.9539455785416

I think the test case is not a very good case. Here it is a pointwise conv whose kernel size is 1x1, and there is no paddings. In this case, actually we can add another shortcut to use batched gemm directly instead of doing that after unfold2d, same for the backward pass.

piyushmh commented:
So I'm a bootcamper interested in joining the pytorch team. This task was to specifically optimize grouped 2d convolutions to batches. Is it okay we work on further optimizations in follow up diffs? I've worked on this task for a week already and would like to wrap it up as soon as I can.

BIT-silence commented:
> So I'm a bootcamper interested in joining the pytorch team. This task was to specifically optimize grouped 2d convolutions to batches. Is it okay we work on further optimizations in follow up diffs? I've worked on this task for a week already and would like to wrap it up as soon as I can.

I thinks actually there should be two tasks.
1. To optimize grouped conv2d by unfold -> bmm
2. To optimize pointwise conv2d by bmm without unfold.

I guess this PR is desired to focus on the 1st one. If so that's totally fine to left the 2nd one as another task in the future. Here my point is, if we want to focus on the 1st one, I'd suggest to benchmark some other more generalized conv2d cases, for example kernel_size = 3x3 and padding = 1. Currently we have put some benchmark for 1x1 grouped conv here, which is not a very common case except ShuffleNet. So I'd suggest to provide more benchmarks here and left some kind of todo for the pointwise conv cases.

piyushmh commented:
> > So I'm a bootcamper interested in joining the pytorch team. This task was to specifically optimize grouped 2d convolutions to batches. Is it okay we work on further optimizations in follow up diffs? I've worked on this task for a week already and would like to wrap it up as soon as I can.
> 
> I think actually there should be two tasks.
> 
> 1. To optimize grouped conv2d by unfold -> bmm
> 2. To optimize pointwise conv2d by bmm without unfold.
> 
> I guess this PR is desired to focus on the 1st one. If so that's totally fine to left the 2nd one as another task in the future. Here my point is, if we want to focus on the 1st one, I'd suggest to benchmark some other more generalized conv2d cases, for example kernel_size = 3x3 and padding = 1. Currently we have put some benchmark for 1x1 grouped conv here, which is not a very common case except ShuffleNet. So I'd suggest to provide more benchmarks here and left some kind of todo for the pointwise conv cases.

Thanks for your detailed reply. I've updated the summary and title to more accurately reflect this. I've also added the script and benchmark numbers for various different combinations in the PR summary.

lly-zero-one commented:
> > > So I'm a bootcamper interested in joining the pytorch team. This task was to specifically optimize grouped 2d convolutions to batches. Is it okay we work on further optimizations in follow up diffs? I've worked on this task for a week already and would like to wrap it up as soon as I can.
> > 
> > 
> > I think actually there should be two tasks.
> > 
> > 1. To optimize grouped conv2d by unfold -> bmm
> > 2. To optimize pointwise conv2d by bmm without unfold.
> > 
> > I guess this PR is desired to focus on the 1st one. If so that's totally fine to left the 2nd one as another task in the future. Here my point is, if we want to focus on the 1st one, I'd suggest to benchmark some other more generalized conv2d cases, for example kernel_size = 3x3 and padding = 1. Currently we have put some benchmark for 1x1 grouped conv here, which is not a very common case except ShuffleNet. So I'd suggest to provide more benchmarks here and left some kind of todo for the pointwise conv cases.
> 
> Thanks for your detailed reply. I've updated the summary and title to more accurately reflect this. I've also added the script and benchmark numbers for various different combinations in the PR summary.

Thanks for working on this. The second optimization is partially done in https://github.com/pytorch/pytorch/pull/40324. Originally, it did not allow the group convolution in slow_conv2d_update_output_frame. We can leave the second one in another task. Also make sure you disable the mkldnn when you ran the test, since we are using mkldnn for group convolution by default.

piyushmh commented:
> > > > So I'm a bootcamper interested in joining the pytorch team. This task was to specifically optimize grouped 2d convolutions to batches. Is it okay we work on further optimizations in follow up diffs? I've worked on this task for a week already and would like to wrap it up as soon as I can.
> > > 
> > > 
> > > I think actually there should be two tasks.
> > > 
> > > 1. To optimize grouped conv2d by unfold -> bmm
> > > 2. To optimize pointwise conv2d by bmm without unfold.
> > > 
> > > I guess this PR is desired to focus on the 1st one. If so that's totally fine to left the 2nd one as another task in the future. Here my point is, if we want to focus on the 1st one, I'd suggest to benchmark some other more generalized conv2d cases, for example kernel_size = 3x3 and padding = 1. Currently we have put some benchmark for 1x1 grouped conv here, which is not a very common case except ShuffleNet. So I'd suggest to provide more benchmarks here and left some kind of todo for the pointwise conv cases.
> > 
> > 
> > Thanks for your detailed reply. I've updated the summary and title to more accurately reflect this. I've also added the script and benchmark numbers for various different combinations in the PR summary.
> 
> Thanks for working on this. The second optimization is partially done in #40324. Originally, it did not allow the group convolution in slow_conv2d_update_output_frame. We can leave the second one in another task. Also make sure you disable the mkldnn when you ran the test, since we are using mkldnn for group convolution by default.

I pasted the script I used in the PR summary. I turn off mkldnn for group convolution using this 
```
torch._C._set_mkldnn_enabled(False)
```

piyushmh commented:
Please let me know if there are any more comments, otherwise I'll go ahead and land this diff.
